### PR TITLE
Replace deprecated KeyWindow property for iOS15

### PR DESCRIPTION
**Pull request by ghost**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Fix intermittent NRE crash for iOS15.1 by having an alternative way to get the window object.

### :arrow_heading_down: What is the current behavior?
When restarting an app using iOS15.1 the pop-up crashes and doesn't show up.

### :new: What is the new behavior (if this is a feature change)?
The pop-up works as expected on iOS 15.1.

### :boom: Does this PR introduce a breaking change?
No

### :bug: Recommendations for testing
Restart the app a few times while opening a pop-up using the plugin.

### :memo: Links to relevant issues/docs
Applied fix from this source:
https://developer.apple.com/forums/thread/695932

### :thinking: Checklist before submitting

- [ ] All projects build
- [x] Follows style guide lines 
- [ ] Relevant documentation was updated
- [x] Rebased onto current develop

### CODE REVIEW

**Comment by ghost** [comment]
@LuckyDucko can you review this PR when you have time, please? Let me know if you need more info regarding this PR. Thanks!

File: Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs [hunk 1]
```diff
@@ -50,8 +50,9 @@ public Task AddAsync(PopupPage page)
 
             page.DescendantRemoved += HandleChildRemoved;
 
-            if (UIApplication.SharedApplication.KeyWindow.WindowLevel == UIWindowLevel.Normal)
-                UIApplication.SharedApplication.KeyWindow.WindowLevel = -1;
+            var uiWindow = UIApplication.SharedApplication.KeyWindow ?? UIApplication.SharedApplication.Delegate.GetWindow();
```
Comment (xleon) [inline Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs]: According to this [SO answer](https://stackoverflow.com/a/58031897/1850606), a better solution to get the window might look like this:
```csharp
public static UIWindow GetKeyWindow(this UIApplication application)
{
    if (!UIDevice.CurrentDevice.CheckSystemVersion(13, 0)) 
        return application.KeyWindow; // deprecated in iOS 13
    
    var window = application
        .ConnectedScenes
        .ToArray()
        .OfType<UIWindowScene>()
        .SelectMany(scene => scene.Windows)
        .FirstOrDefault(window => window.IsKeyWindow);

    return window;
}
```

Comment (xleon) [reply Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs]: I've just realised there are similar code in this very same class

Comment (ghost) [reply Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs]: Thanks! I've updated the code to reflect that.

File: Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs [hunk 2]
```diff
@@ -97,5 +97,20 @@ public static void UpdateSize(this PopupPageRenderer renderer)
             else if (needForceLayout)
                 currentElement.ForceLayout();
         }
+
+        public static UIWindow GetKeyWindow(this UIApplication application)
+        {
+            if (!UIDevice.CurrentDevice.CheckSystemVersion(13, 0))
```
Comment (LuckyDucko) [inline Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs]: Could you use the `IsiOS13OrNewer` bool in this class

File: Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs [hunk 3]
```diff
@@ -97,5 +97,20 @@ public static void UpdateSize(this PopupPageRenderer renderer)
...
+                return UIApplication.SharedApplication.KeyWindow;
+
+            var window = application
```
Comment (LuckyDucko) [inline Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs:108]: i noticed in the add there is mention of

```csharp
 if (UIApplication.SharedApplication.ConnectedScenes.ToArray()
                    .FirstOrDefault(x => x.ActivationState == UISceneActivationState.ForegroundActive && x is UIWindowScene) is UIWindowScene connectedScene)
```

your version here thanks to @xleon seems to be much cleaner than this. Just tagging this for a future consideration.

**Review by LuckyDucko — CHANGES_REQUESTED** [review]
Good stuff, great catch on the deprecated functionality.

Could you just change to use the `IsiOS13OrNewer`, otherwise good to go 👍🏻

File: Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs (hunk 3 above)
Comment (xleon) [reply Rg.Plugins.Popup/Platforms/Ios/Extensions/PlatformExtension.cs:108]: What I´m not sure is if we need this part (which my code lacks): 
`x.ActivationState == UISceneActivationState.ForegroundActive`. 
What I understood from the SO thread is that we don't. But I need to test it.

File: Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs [hunk 4]
```diff
@@ -50,8 +50,9 @@ public Task AddAsync(PopupPage page)
 
             page.DescendantRemoved += HandleChildRemoved;
 
-            if (UIApplication.SharedApplication.KeyWindow.WindowLevel == UIWindowLevel.Normal)
-                UIApplication.SharedApplication.KeyWindow.WindowLevel = -1;
+            var keyWindow = UIApplication.SharedApplication.GetKeyWindow();
+            if (keyWindow.WindowLevel == UIWindowLevel.Normal)
```
Comment (xleon) [inline Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs]: `if (keyWindow?.WindowLevel == UIWindowLevel.Normal)`
keyWindow might be null

File: Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs [hunk 5]
```diff
@@ -115,10 +116,11 @@ public async Task RemoveAsync(PopupPage page)
                     window = null;
                 }
 
-                if(_windows.Count > 0)
+                var keyWindow = UIApplication.SharedApplication.GetKeyWindow();
+                if (_windows.Count > 0)
                     _windows.Last().WindowLevel = UIWindowLevel.Normal;
-                else if (UIApplication.SharedApplication.KeyWindow.WindowLevel == -1)
-                    UIApplication.SharedApplication.KeyWindow.WindowLevel = UIWindowLevel.Normal;
+                else if (keyWindow.WindowLevel == -1)
```
Comment (xleon) [inline Rg.Plugins.Popup/Platforms/Ios/Impl/PopupPlatformIos.cs]: `if (keyWindow?.WindowLevel == UIWindowLevel.Normal)`
keyWindow might be null

**Comment by xleon** [comment]
Hi guys, can we please merge this PR?
The requested changes have been done already by @arctouch-andresato

**Comment by LuckyDucko** [comment]
@xleon My Apologies, usually i leave @martijn00 for merging, as he is in control of updating the NuGet. However, as i have reviewed this, i'm able to merge, just not able to make a NuGet

**Comment by ghost** [comment]
Hi @rotorgames, @martijn00, do you know when we'll have a new Nuget release with this change?

**Comment by ramonB1996** [comment]
Yes please, this bug causes problems for the `MaterialMenuButton` on iOS 15+ as well, so would be nice if the nuget package is coming soon! :)

**Comment by rjantz2** [comment]
@martijn00 Could we get an updated nuget package soon, please?

**Comment by xleon** [comment]
This fix is actually working. Thanks for the help.

**Pull request merged**